### PR TITLE
Clarify space partitioning due to multi-node

### PR DESCRIPTION
**Pull request by k-rus**
Space partitioning is rare case for non-distributed hypertables, but is
important for multi-node to be effective. This commit clarifies this.

### CODE REVIEW

File: api.md [hunk 1]
```diff
@@ -82,13 +82,29 @@ partitioning (e.g., for a second time partition) or hash partitioning.
 converted to a hypertable (via `create_hypertable`), but must similarly
 be run only on an empty hypertable.
 
-**Space partitions**: The use of additional partitioning is a very
-specialized use case.  **Most users will not need to use it.**
+**Space partitions**: Using space partitions is important for efficient scale out 
+in [multi-node setup][multinode]. In the case of 
+single-node the use of additional partitioning is a very
+specialized use case and most users will not need it.
```
Comment (mfreed) [inline api.md:88]:
```suggestion
**Space partitions**: Using space partitions is highly recommended
for [distributed hypertables](#create_distributed_hypertable) to achieve
efficient scale-out performance. For [regular hypertables](#create_hypertable)
that exist only a single node, the use of additional partitioning is a
specialized use case and not recommended for most users.
```

File: api.md [hunk 2]
```diff
@@ -82,13 +82,29 @@ partitioning (e.g., for a second time partition) or hash partitioning.
...
 
 Space partitions use hashing: Every distinct item is hashed to one of
 *N* buckets.  Remember that we are already using (flexible) time
 intervals to manage chunk sizes; the main purpose of space
-partitioning is to enable parallel I/O to the same time interval.
+partitioning is to enable parallel utilization of multiple data nodes or 
+parallel I/O on a single node to the same time interval.
+
+### Space partitions for multi-node
```
Comment (mfreed) [inline api.md:96]:
```suggestion
### Parallelizing queries across multiple data nodes [](add_dimension-multi_node)
```

File: api.md [hunk 3]
```diff
@@ -82,13 +82,29 @@ partitioning (e.g., for a second time partition) or hash partitioning.
...
+Space partitioning allows to parallelize inserts into different data nodes
+and increase the ingesting rate. Select queries will also benefit when 
+operations, which reduce data, e.g., conditions in WHERE clause or
+aggregate operators, are pushed to data nodes. However, some operations 
+are not pushed, e.g., joins or non-parallelizable aggregates, and thus
+all needed data will be transferred to the access node, which might be expensive.
+
+It is recommended to set the number of partitions to the number of attached data nodes.
+
+### Parallel I/O on a single node
```
Comment (mfreed) [inline api.md:107]:
```suggestion
### Parallelizing disk I/O on a single node [](add_dimension-single_node)
```

File: api.md [hunk 4]
```diff
@@ -82,13 +82,29 @@ partitioning (e.g., for a second time partition) or hash partitioning.
...
+in [multi-node setup][multinode]. In the case of 
+single-node the use of additional partitioning is a very
+specialized use case and most users will not need it.
 
 Space partitions use hashing: Every distinct item is hashed to one of
 *N* buckets.  Remember that we are already using (flexible) time
 intervals to manage chunk sizes; the main purpose of space
-partitioning is to enable parallel I/O to the same time interval.
+partitioning is to enable parallel utilization of multiple data nodes or 
+parallel I/O on a single node to the same time interval.
```
Comment (mfreed) [inline api.md:94]:
```suggestion
partitioning is to enable parallelization across multiple 
data nodes (in the case of distributed hypertables) or
across multiple disks within the same time interval
(in the case of single-node deployments).
```

File: api.md [hunk 5]
```diff
@@ -187,6 +202,16 @@ SELECT add_dimension('conditions', 'device_id', number_partitions => 2);
 SELECT add_dimension('conditions', 'device_id', number_partitions => 2, if_not_exists => true);
 ```
 
+First, create a multinode with two data nodes, then convert table `conditions` to 
+a distributed hypertable with just time partitioning on column `time`, 
+finally add a space partitioning on `location` with two partitions:
```
Comment (mfreed) [inline api.md:207]:
```suggestion
```suggestion
Now in a multi-node example for distributed hypertables, 
setup your cluster to have one access node and two data nodes, 
and configure your access node for access to the data nodes.
Then, convert your table `conditions` to 
a distributed hypertable with just time partitioning on column `time`, 
and finally add a space partitioning dimension on `location`.  
By default, the `number_partitions` will be set to the number of data nodes
in the distributed hypertable `conditions`.
```

File: api.md [hunk 6]
```diff
@@ -187,6 +202,16 @@ SELECT add_dimension('conditions', 'device_id', number_partitions => 2);
...
+```sql
+SELECT add_data_node('data_node1', host => 'node1_host_address');
+SELECT add_data_node('data_node2', host => 'node2_host_address');
+SELECT create_distributed_hypertable('conditions', 'time');
+SELECT add_dimension('conditions', 'location', number_partitions => 2);
```
Comment (mfreed) [inline api.md:212]:
```suggestion
SELECT add_dimension('conditions', 'location');
```

File: api.md [hunk 7]
```diff
@@ -82,13 +82,29 @@ partitioning (e.g., for a second time partition) or hash partitioning.
...
+### Space partitions for multi-node
+
+Space partitioning allows to parallelize inserts into different data nodes
+and increase the ingesting rate. Select queries will also benefit when 
+operations, which reduce data, e.g., conditions in WHERE clause or
+aggregate operators, are pushed to data nodes. However, some operations 
+are not pushed, e.g., joins or non-parallelizable aggregates, and thus
+all needed data will be transferred to the access node, which might be expensive.
+
+It is recommended to set the number of partitions to the number of attached data nodes.
```
Comment (mfreed) [inline api.md:105]:
```suggestion
In a distributed hypertable, space partitioning enables inserts to be
parallelized across data nodes, even while the inserted rows share
timestamps from the same time interval, and thus increases the ingest rate.
Query performance also benefits by being able to parallelize queries 
across nodes, particularly when full or partial aggregations can be
"pushed down" to data nodes (e.g., as in the query
`avg(temperature) FROM conditions GROUP BY hour, location`
when using `location` as a space partition). Please see our
[best practices about partitioning in distributed hypertables][distributed-hypertable-partitioning-best-practices]
for more information.
```

File: api.md [hunk 8]
```diff
@@ -3669,6 +3697,7 @@ and then inspect `dump_file.txt` before sending it together with a bug report or
 [postgres-lock]: https://www.postgresql.org/docs/current/sql-lock.html
 [migrate-from-postgresql]: /getting-started/migrating-data
 [memory-units]: https://www.postgresql.org/docs/current/static/config-setting.html#CONFIG-SETTING-NAMES-VALUES
+[multinode]: /getting-started/setup-multi-node-basic
```
Comment (mfreed) [inline api.md:3700]:
```suggestion
[multinode]: /getting-started/setup-multi-node-basic
[distributed-hypertable-partitioning-best-practices]: /using-timescaledb/distributed-hypertables#partitioning-best-practices
```

File: api.md (hunk 6 above)
Comment (k-rus) [reply api.md:217]: @mfreed It doesn't work in this way. `add_dimension` in you suggestion will return an error that `number_partitions` (or time interval) must be provided.

Comment (k-rus) [reply api.md:217]: The default exists only for `create_distributed_hypertable`.

Comment (mfreed) [reply api.md:217]: Sounds good.  Curious -> is this fundamental or just an API thing we didn't get around to updating?

Comment (k-rus) [reply api.md:221]: Being discussed outside the PR.

File: api.md [hunk 9]
```diff
@@ -82,13 +82,34 @@ partitioning (e.g., for a second time partition) or hash partitioning.
 converted to a hypertable (via `create_hypertable`), but must similarly
 be run only on an empty hypertable.
 
-**Space partitions**: The use of additional partitioning is a very
-specialized use case.  **Most users will not need to use it.**
+**Space partitions**: Using space partitions is highly recommended
+for [distributed hypertables](#create_distributed_hypertable) to achieve
+efficient scale-out performance. For [regular hypertables](#create_hypertable)
+that exist only a single node, the use of additional partitioning is a
```
Comment (pmwkaa) [inline api.md:88]: only on a?

File: api.md [hunk 10]
```diff
@@ -82,13 +82,34 @@ partitioning (e.g., for a second time partition) or hash partitioning.
...
 Space partitions use hashing: Every distinct item is hashed to one of
 *N* buckets.  Remember that we are already using (flexible) time
 intervals to manage chunk sizes; the main purpose of space
-partitioning is to enable parallel I/O to the same time interval.
+partitioning is to enable parallelization across multiple 
+data nodes (in the case of distributed hypertables) or
+across multiple disks within the same time interval
+(in the case of single-node deployments).
+
+### Parallelizing queries across multiple data nodes [](add_dimension-multi_node)
```
Comment (pmwkaa) [inline api.md:99]: Curious if we should move this sections somewhere else, like performance or optimization

File: api.md [hunk 11]
```diff
@@ -187,6 +207,20 @@ SELECT add_dimension('conditions', 'device_id', number_partitions => 2);
...
+Now in a multi-node example for distributed hypertables with a cluster 
+of one access node and two data nodes, configure the access node for 
+access to the two data nodes. Then, convert table `conditions` to 
+a distributed hypertable with just time partitioning on column `time`, 
+and finally add a space partitioning dimension on `location`
+with two partitions (as the number of the attached data nodes).
+
+```sql
+SELECT add_data_node('dn1', host => 'dn1.example.com');
+SELECT add_data_node('dn2', host => 'dn1.example.com');
```
Comment (pmwkaa) [inline api.md:219]: `dn2` is using `dn1.example.com` host

**Review by pmwkaa — APPROVED** [review]

File: api.md (hunk 11 above)
Comment (k-rus) [reply api.md]: Thank you! Fixed.

File: api.md (hunk 9 above)
Comment (k-rus) [reply api.md]: I improved this place.

File: api.md (hunk 10 above)
Comment (k-rus) [reply api.md:99]: I suggest to discuss and consider this outside this PR.

**Review by mfreed — APPROVED** [review]

**Pull request merged**